tests/integration/setup_environment.py

```diff
@@ -0,0 +1,99 @@
+    """Try to find a free port in the given range."""
+    for port in range(start, end):
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            if s.connect_ex(("127.0.0.1", port)) != 0:  # Port is free
+                return port
+    raise DockerContainerError(f"No free ports found in the range {start} to {end}")
+
+
+class ADCMPostgresContainer(PostgresContainer):
+    url: str
+
+    def __init__(self: Self, image: str) -> None:
+        super().__init__(image)
+
+    def setup_postgres(self: Self, username: str, password: str, adcm_db_name: str, container_network: Network) -> None:
+        postgres_port = find_free_port(postgres_port_range[0], postgres_port_range[1])
+        self.with_name(f"{postgres_name}_{postgres_port}")
+        self.with_bind_ports(postgres_port, postgres_port)
+        self.password = password
+        self.with_network(container_network)
+
+        self.start()
+        wait_container_is_ready(self)
+
+        init_db_cmds = (
+            f"CREATE USER {username} WITH ENCRYPTED PASSWORD '{password}'; "
+            f"CREATE DATABASE {adcm_db_name} OWNER {username};"
+        )
+        self.with_command(
+            f"PGPASSWORD={self.password} psql --username {self.username} --host 127.0.0.1 -c '{init_db_cmds}'"
+        )
+
+        wait_for_logs(self, "database system is ready to accept connections")
+
+        self.url = f"http://{self.get_container_host_ip()}:{self.get_exposed_port(postgres_port)}"
+
+
+class ADCMContainer(DockerContainer):
+    url: str
+
+    def __init__(self: Self, image: str) -> None:
+        super().__init__(image)
+
+    def setup_container(self: Self, container_network: Network) -> None:
+        adcm_port = find_free_port(adcm_port_range[0], adcm_port_range[1])
+        self.with_name(f"{adcm_container_name}_{adcm_port}")
+        self.with_network(container_network)
+        self.with_bind_ports(adcm_port, adcm_port)
+
+        self.start()
+
+        wait_container_is_ready(self)
+        wait_for_logs(self, "Run Nginx ...")
+
+        self.url = f"http://{self.get_container_host_ip()}:{self.get_exposed_port(adcm_port)}"
+
+
+class SetupManager:
+    def __init__(self: Self, postgres_image: str, adcm_image: str) -> None:
+        self.network = Network()
+        self.postgres_container = ADCMPostgresContainer(postgres_image)
+        self.adcm_container = ADCMContainer(adcm_image)
+
+    def __enter__(self: Self) -> Self:
+        # Start the network
+        self.network.__enter__()
+
+        self.postgres_container.setup_postgres(db_user, db_password, db_name, self.network)
+        self.adcm_container.setup_container(self.network)
+
+        return self
+
+    def __exit__(self: Self, exc_type, exc_val, exc_tb) -> None:  # noqa: ANN001
+        # Clean up Docker containers and network
+        self.adcm_container.stop()
+        self.postgres_container.stop()
+        self.network.__exit__(exc_type, exc_val, exc_tb)
```

Comment:
since we use pytest, I think we should use fixtures for such preparation 

Fixtures to write and their description:
1. network - yield network; scope seasion
2. postgres - yield postgres container; scope function for now
3. adcm - yield adcm container; scope function

example:
```python
@pytest.fixture()
def postgres(network):
    with Postgres() as pg:
        yield pg
```

add typehints, use your Postgres and ADCM implementations, make fixture hierarchy correct

Reply:
done